fractional-NFT/project/token-distributor-contract/tests/functions/buyback.rs

```diff
@@ -0,0 +1,429 @@
+    #[should_panic(expected = "DistributionDoesNotExist")]
+    async fn when_token_distribution_does_not_exist() {
+        let (
+            _deployer,
+            owner1,
+            _owner2,
+            _token_distributor_contract,
+            fractional_nft_contract,
+            _nft_contract,
+            asset_contract,
+        ) = setup().await;
+        let (_reserve_price, token_price, _token_supply, purchase_amount, asset_supply) =
+            defaults().await;
+
+        mint_and_send_to_address(
+            asset_supply,
+            &owner1.asset,
+            Address::new(*owner1.wallet.address().hash()),
+        )
+        .await;
+
+        buyback(
+            purchase_amount * token_price,
+            &owner1.token_distributor,
+            asset_contract.clone(),
+            fractional_nft_contract.clone(),
+            token_price,
+        )
+        .await;
+    }
+
+    #[tokio::test]
+    #[should_panic(expected = "NotTokenAdmin")]
+    async fn when_no_owner() {
+        let (
+            _deployer,
+            owner1,
+            owner2,
+            _token_distributor_contract,
+            fractional_nft_contract,
+            nft_contract,
+            asset_contract,
+        ) = setup().await;
+        let (reserve_price, token_price, token_supply, purchase_amount, asset_supply) =
+            defaults().await;
+
+        let owner_identity = Identity::Address(owner1.wallet.address().into());
+        let fractional_nft_identity = Identity::ContractId(fractional_nft_contract.into());
+
+        mint(1, &owner1.nft, owner_identity.clone()).await;
+        approve(Some(fractional_nft_identity.clone()), &owner1.nft, 0).await;
+        create(
+            &owner1.token_distributor,
+            asset_contract.clone(),
+            fractional_nft_contract.clone(),
+            nft_contract.clone(),
+            Some(reserve_price),
+            None,
+            token_price,
+            token_supply,
+            0,
+        )
+        .await;
+        mint_and_send_to_address(
+            asset_supply,
+            &owner2.asset,
+            Address::new(*owner2.wallet.address().hash()),
+        )
+        .await;
+        mint_and_send_to_address(
+            asset_supply,
+            &owner1.asset,
+            Address::new(*owner1.wallet.address().hash()),
+        )
+        .await;
+        purchase(
+            purchase_amount,
+            &owner2.token_distributor,
+            asset_contract.clone(),
+            fractional_nft_contract.clone(),
+            token_price,
+        )
+        .await;
+
+        buyback(
+            purchase_amount * token_price,
+            &owner1.token_distributor,
+            asset_contract.clone(),
+            fractional_nft_contract.clone(),
+            token_price,
+        )
+        .await;
+    }
+
+    #[tokio::test]
+    #[should_panic(expected = "InvalidAssetTransfer")]
+    async fn when_incorrect_amount_provided() {
```

Comment:
This requirement occurs after the `InvalidState` check. Can this test be moved below that one so that reading the cases in the `main.sw` appear to be in the same order as the tests? It's easier to see if something is or is not missing that way.